src/Stratis.Bitcoin.Features.PoA/Voting/VotingManager.cs

```diff
@@ -209,14 +222,74 @@ private void onBlockConnected(ChainedHeaderBlock chBlock)
+        public void onBlockDisconnected(ChainedHeaderBlock chBlock)
         {
-            // TODO it should handle reorgs and revert votes that were applied.
+            byte[] rawVotingData = this.votingDataEncoder.ExtractRawVotingData(chBlock.Block.Transactions[0]);
+
+            List<VotingData> votingDataList = this.votingDataEncoder.Decode(rawVotingData);
+            votingDataList.Reverse();
+
+            if (rawVotingData == null)
+            {
+                this.logger.LogTrace("(-)[NO_VOTING_DATA]");
+                return;
+            }
+
+            lock (this.locker)
+            {
+                foreach (VotingData votingData in votingDataList)
+                {
+                    // Poll that was finished in the block being disconnected.
+                    Poll pollToRevert = this.finishedPolls.SingleOrDefault(x => x.PollAppliedBlockHash == chBlock.Block.GetHash() && x.VotingData == votingData);
+
+                    if (pollToRevert != null)
+                    {
+                        this.pollResultExecutor.RevertChange(pollToRevert.VotingData);
+
+                        pollToRevert.PollAppliedBlockHash = null;
+
+                        this.finishedPolls.Remove(pollToRevert);
+                        this.pendingPolls.Add(pollToRevert);
+
+                        this.pendingPolls = this.pendingPolls.OrderBy(x => x.Id).ToList();
+                    }
+
+                    // Pub key of a fed member that created voting data.
+                    string fedMemberKeyHex = this.slotsManager.GetPubKeyForTimestamp(chBlock.Block.Header.Time).ToHex();
+
+                    Poll targetPendingPoll = this.pendingPolls.Single(x => x.VotingData == votingData);
+
+                    targetPendingPoll.PubKeysHexVotedInFavor.Remove(fedMemberKeyHex);
+
+                    if (targetPendingPoll.PubKeysHexVotedInFavor.Count == 0)
+                        this.pendingPolls.Remove(targetPendingPoll);
+                }
+
+                this.SavePolls(this.pendingPolls, PendingPollsDbKey);
+                this.SavePolls(this.finishedPolls, FinishedPollsDbKey);
+            }
         }
 
-        // TODO tests
-        // add tests that will check reorg that adds or removes fed members
-        // test that vote of a fed member that is no longer there is not active anymore
-        // test case when we have 2 votes in a block and because 1 is executed before the other other no longer has enough votes
+        /// <summary>Provides id for the next poll that is to be created.</summary>
+        /// <remarks>Should be locked by <see cref="locker"/>.</remarks>
+        private int GetNextPollIdLocked()
+        {
+            int largestId = -1;
+
+            foreach (Poll finishedPoll in this.finishedPolls)
+            {
+                if (finishedPoll.Id > largestId)
+                    largestId = finishedPoll.Id;
+            }
+
+            foreach (Poll pendingPoll in this.pendingPolls)
```

Comment:
This means after a reorg a pool id will be reused right? which I think is correct

Reply:
yes. order will be restored and we need IDs for it.